subsys/bluetooth/host/hci_core.c

```diff
@@ -1941,15 +1941,21 @@ static void conn_auto_initiate(struct bt_conn *conn)
 	}
 
 	if (IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE) &&
-	    BT_FEAT_LE_DLE(bt_dev.le.features)) {
-		uint16_t tx_octets, tx_time;
+		BT_FEAT_LE_DLE(bt_dev.le.features)) {
+		if (bt_dev.drv->quirks & BT_QUIRK_NO_AUTO_DLE) {
+			uint16_t tx_octets, tx_time;
 
-		err = hci_le_read_max_data_len(&tx_octets, &tx_time);
-		if (!err) {
-			err = bt_le_set_data_len(conn, tx_octets, tx_time);
-			if (err) {
-				BT_ERR("Failed to set data len (%d)", err);
+			err = hci_le_read_max_data_len(&tx_octets, &tx_time);
+			if (!err) {
+				err = bt_le_set_data_len(conn, tx_octets, tx_time);
+				if (err) {
+					BT_ERR("Failed to set data len (%d)", err);
+				}
 			}
+		} else {
+			/* No need to auto-initiate DLE procedure.
+			 * It is done by the controller. */
+			atomic_set_bit(conn->flags, BT_CONN_AUTO_DATA_LEN_COMPLETE);
```

Comment:
Wouldn't it be enough to only set this in the le_data_len_change?
Now the application can start the user initiated data length update procedure while the controller initiated data length update might be running.